logics/impls/marketplace/marketplace_sale.rs

```diff
@@ -170,66 +170,61 @@ where
-                    ink_prelude::vec::Vec::new(),
-                ) == Ok(())
-                {
-                    Self::env()
-                        .transfer(token_owner, seller_fee)
-                        .map_err(|_| MarketplaceError::TransferToOwnerFailed)?;
-                    Self::env()
-                        .transfer(self.data::<Data>().market_fee_recipient, marketplace_fee)
-                        .map_err(|_| MarketplaceError::TransferToMarketplaceFailed)?;
-                    Self::env()
-                        .transfer(collection.royalty_receiver, author_royalty)
-                        .map_err(|_| MarketplaceError::TransferToAuthorFailed)?;
-
-                    return Ok(())
-                } else {
-                    return Err(MarketplaceError::UnableToTransferToken)
-                }
-            } else {
-                return Err(MarketplaceError::NotOwner)
+            .ok_or(MarketplaceError::ItemNotListedForSale)?;
+
+        let token_owner = PSP34Ref::owner_of(&contract_address, token_id.clone())
+            .ok_or(MarketplaceError::CanNotDetermineOwner)?;
+
+        let value = Self::env().transferred_value();
+        self.check_value(value, item.price)?;
+
+        let collection = self
+            .data::<Data>()
+            .registered_contracts
+            .get(&contract_address)
+            .ok_or(MarketplaceError::NotRegisteredContract)?;
+
+        let marketplace_fee = value
+            .checked_mul(self.data::<Data>().fee as u128)
+            .unwrap_or_default()
+            / 10_000;
+        let author_royalty = value
+            .checked_mul(collection.royalty as u128)
+            .unwrap_or_default()
+            / 10_000;
+        let seller_fee = value
+            .checked_sub(marketplace_fee)
+            .unwrap_or_default()
+            .checked_sub(author_royalty)
+            .unwrap_or_default();
+
+        let caller = Self::env().caller();
+        ensure!(token_owner != caller, MarketplaceError::AlreadyOwner);
+
+        match PSP34Ref::transfer(
+            &contract_address,
+            caller,
+            token_id,
+            ink_prelude::vec::Vec::new(),
+        ) {
+            Ok(()) => {
+                Self::env()
+                    .transfer(token_owner, seller_fee)
+                    .map_err(|_| MarketplaceError::TransferToOwnerFailed)?;
+                Self::env()
+                    .transfer(self.data::<Data>().market_fee_recipient, marketplace_fee)
+                    .map_err(|_| MarketplaceError::TransferToMarketplaceFailed)?;
+                Self::env()
+                    .transfer(collection.royalty_receiver, author_royalty)
+                    .map_err(|_| MarketplaceError::TransferToAuthorFailed)?;
+
+                return Ok(())
             }
-        } else {
-            return Err(MarketplaceError::ItemNotListedForSale)
+            _ => return Err(MarketplaceError::UnableToTransferToken),
```

Comment:
PSP34Ref::transfer returns a Result<T, E> whivh is an enum with two variant: Ok & Err
So here handle `Err` case and not `_`
Also you can add return before `match PSP34Ref::transfer(`
and remove it in the match branchs